amq-quarkus-school-timetabling/amq-quarkus-school-timetabling-solver/src/main/java/org/acme/schooltimetabling/messaging/TimeTableMessagingHandler.java

```
@@ -0,0 +1,109 @@
+import javax.inject.Inject;
+
+import org.acme.schooltimetabling.domain.TimeTable;
+import org.acme.schooltimetabling.message.SolverRequest;
+import org.acme.schooltimetabling.message.SolverResponse;
+import org.eclipse.microprofile.reactive.messaging.Channel;
+import org.eclipse.microprofile.reactive.messaging.Emitter;
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.api.solver.SolverFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+@ApplicationScoped
+public class TimeTableMessagingHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TimeTableMessagingHandler.class);
+    public static final String SOLVER_REQUEST_CHANNEL = "solver_request";
+    public static final String SOLVER_RESPONSE_CHANNEL = "solver_response";
+
+    Solver<TimeTable> solver;
+
+    @Inject
+    ObjectMapper objectMapper;
+
+    @Inject
+    @Channel(SOLVER_RESPONSE_CHANNEL)
+    Emitter<String> solverResponseEmitter;
+
+    @Inject
+    TimeTableMessagingHandler(SolverFactory<TimeTable> solverFactory) {
+        solver = solverFactory.buildSolver();
+    }
+
+    @Incoming(SOLVER_REQUEST_CHANNEL)
+    public CompletionStage<Void> solve(Message<String> solverRequestMessage) {
+        return CompletableFuture.runAsync(() -> {
+            SolverRequest solverRequest = null;
+            try {
+                solverRequest = objectMapper.readValue(solverRequestMessage.getPayload(), SolverRequest.class);
+            } catch (JsonProcessingException ex) {
+                LOGGER.warn("Unable to deserialize solver request from JSON.", ex);
+                // Bad request should go directly to a DLQ.
+                solverRequestMessage.nack(ex);
+                return;
```

Normally, when things go wrong, we throw an exception instead of doing "warn" and returning, as this can lead to the exception being ignored. Especially as JsonProcessingException is - unlike a "wrong password" exception - something that shouldn't happen normally.

If there's a good reason why we can't just throw an exception (it brings the whole system down - and that would be a bad thing - usually that is a good thing so the bug get fixed), so the warn makes sense, let's add a comment why this case is exceptional.

---

Ok, reading further it makes 100% sense why we don't bring down the whole system here, as the client makes a mistake. But I would except that if you do throw an exception here, that quarkus won't bring down the whole system.
Just like the swing event thread and http servlet thread have a "don't-kill-the-thread" general exception handler, I 'd expect the quarkus reactive thing to have this too. Something to try out.

If that's there, we can nack the message and throw the exception, without the log line.
Why is that better? Because any line could throw an exception.  `objectMapper.readValue` could throw an exception that is not caught by JsonProcessingException. solver.solve could throw an Error that is not caught by Exception.

---

The exception handling you're suggesting works only with automatic acknowledgment. Here we use manual acknowledgment via the `Message` class and thus we are responsible for calling either `ack()` or `nack()`.